main.h

```diff
@@ -0,0 +1,10 @@
+#ifndef MAIN_H
+#define MAIN_H
+
+#include <pdf.h>
+
+typedef const char cchar;
+
+void save_to_png(size_t max_store, int page_number, char *pdf, char *output);
```

Comment:
LazyRaster was put in place as a performance project so that it would render into an in-memory buffer on demand. Rendering and saving to a PNG file on the disk goes against this design. (DPE service can already save to PNG and that was the old architecture)
Can you explain the motivations behind this change?

Reply:
This was the quickest/easiest way of connecting MuPDF and the Go runtime. This also exchanges latency and CPU for a lower memory consumption as it does not require the service to hold the file in memory while it's doing the process.

Reply:
I know the cloud team at the time did this whole mupdf/lazyraster project for sake of instant in-memory rendering and decided memory trade-offs are necessary. They deliberately walked away from DPE, which was doing (and still can do) what you propose here - render pages to files as PNG. That was deemed too slow and unresponsive at the time. You're probably trading sub 1ms render time to 80 ms here by hitting the disk and that may be unacceptable. As I said I'm not privy to original metrics and measurements, I wasn't in that team but I know this is the sole reason lazyraster exists.

Reply:
For sure it's not slow as you're expecting, I just did a benchmark creating a file, writing the content coming from a HTTP request to it, reading the content from the file, and then closing and removing it. 

```
goos: darwin
goarch: amd64
pkg: github.com/nitro/lazypdf/v2
cpu: Intel(R) Core(TM) i9-9980HK CPU @ 2.40GHz
BenchmarkFile-16    	     144	   8243701 ns/op	 1583235 B/op	      17 allocs/op
PASS
```

This introduces 8ms on a 1.5MB file. But let me take a look again at how to implement a proper buffer between C and Go. Regards to the C code, did you found anything else? Leaks, etc...?

Reply:
To address some of your comments:
1. The original code did pass a buffer from C to Go. 
2. In terms of loading a 100Mb file to memory - the original PDF file has to be processed by muPDF and the loading of file into memory won't change by doing save to PNG. If you meant to say you want to avoid holding the page render bitmap in memory: in this case you're just pushing the memory use inside mupdf, which will still have to render to a bitmap internally before saving to a PNG. Did you verify how much memory savings do you get by doing this?
3. Most importantly: if you think saving to files doesn't hamper performance too much best thing to do would be to remove lazyraster service all together and just use DPE to render to PNGs. This would also have the added benefit of solving all rendering related inconsistency bugs.
That's why I'd like to discuss this with you over a chat.

Reply:
I think the old version is using the filesystem too, [see here](https://github.com/Nitro/lazypdf/blob/master/faster_raster.go#L259). But anyway, I did change the code to use everything in memory, it's even faster now.

If possible take a look at the C code. I did not found any leak, I left the application running in a loop over several minutes and the memory usage was the same, around 20mb.

Reply:
I'll mark this as solved. Feel free to open again.